langchain/src/util/azure.ts

```diff
@@ -0,0 +1,65 @@
+export interface OpenAIEndpointConfig {
+  azureOpenAIApiDeploymentName?: string;
+  azureOpenAIApiInstanceName?: string;
+  azureOpenAIApiKey?: string;
+  azureOpenAiBasePath?: string;
+  basePath?: string;
+}
+
+/**
+ * This function generates an endpoint URL for (Azure) OpenAI
+ * based on the configuration parameters provided.
+ *
+ * @param {OpenAIEndpointConfig} config - The configuration object for the (Azure) endpoint.
+ *
+ * @property {string} config.azureOpenAIApiDeploymentName - The deployment name of Azure OpenAI.
+ * @property {string} config.azureOpenAIApiInstanceName - The instance name of Azure OpenAI.
+ * @property {string} config.azureOpenAIApiKey - The API Key for Azure OpenAI.
+ * @property {string} config.azureOpenAiBasePath - The base path for Azure OpenAI.
+ * @property {string} config.basePath - The base path URL.
+ *
+ * The function operates as follows:
+ * - If both `azureOpenAiBasePath` and `azureOpenAIApiDeploymentName` are provided, it returns an URL combining these two parameters.
+ * - If `azureOpenAIApiKey` is provided, it checks for `azureOpenAIApiInstanceName` and `azureOpenAIApiDeploymentName` and throws an error if any of these is missing. If both are provided, it generates an URL incorporating these parameters.
+ * - If none of the above conditions are met and `basePath` is not provided, it throws an error.
+ * - The function returns the generated URL as a string.
+ *
+ * @throws Will throw an error if the necessary parameters for generating the URL are missing.
+ *
+ * @returns {string} The generated (Azure) OpenAI endpoint URL.
+ */
+export function getEndpoint(config: OpenAIEndpointConfig): string {
+  const {
+    azureOpenAIApiDeploymentName,
+    azureOpenAIApiInstanceName,
+    azureOpenAIApiKey,
+    azureOpenAiBasePath,
+    basePath,
+  } = config;
+
+  if (azureOpenAiBasePath && azureOpenAIApiDeploymentName) {
+    return `${azureOpenAiBasePath}/openai/deployments/${azureOpenAIApiDeploymentName}`;
```

Comment:
At this point would we rather just allow a custom `basePath` parameter to override the Azure case as well (like it does for the classic OpenAI case)?

I could see a world where Azure changes the `/openai/deployments/` path for some enterprise segment in the future.

Reply:
@jacoblee93 thanks for pointing his out. Would you prefer to return `${azureOpenAiBasePath}/${azureOpenAIApiDeploymentName}` or just `${azureOpenAiBasePath}`?

Reply:
I feel like just overriding the whole thing would make more sesne since that's what `configuration.basePath` does?

Reply:
Please suggest some code if further updates are required before merging

Reply:
I think the above with just `azureOpenAIBasePath` would make the most sense here if you agree?

Reply:
Actually we need different endpoints for the embedding (see https://github.com/hwchase17/langchainjs/blob/main/langchain/src/embeddings/openai.ts#L155) and the llm (see https://github.com/hwchase17/langchainjs/blob/main/langchain/src/llms/openai.ts#L429)

e.g. `AZURE_OPENAI_API_EMBEDDINGS_DEPLOYMENT_NAME='text-embedding-ada-002'` vs `AZURE_OPENAI_API_INSTANCE_NAME='custom-openai-deployment'`

So then we would need at least two variables like `azureOpenAILLMPath` and `azureOpenAIEmbeddingPath` 🤔

Reply:
I decided to go with `return `${azureOpenAIBasePath}/${azureOpenAIApiDeploymentName}`;`, so that `AZURE_OPENAI_API_EMBEDDINGS_DEPLOYMENT_NAME` from https://github.com/hwchase17/langchainjs/blob/4c565af393b4e94d48700334f271e6c6de1eedc0/langchain/src/embeddings/openai.ts#L90 is still used and applied when calling `getEndpoint` in https://github.com/hwchase17/langchainjs/blob/4c565af393b4e94d48700334f271e6c6de1eedc0/langchain/src/embeddings/openai.ts#L169C24-L169C35. I hope that makes in mergable. 😃 🤞
What do you think @jacoblee93 ?

Reply:
Ah ok, fine with me then.